src/sparse_ci/sparse_operator.cc

```diff
@@ -0,0 +1,140 @@
+/*
+ * @BEGIN LICENSE
+ *
+ * Forte: an open-source plugin to Psi4 (https://github.com/psi4/psi4)
+ * that implements a variety of quantum chemistry methods for strongly
+ * correlated electrons.
+ *
+ * Copyright (c) 2012-2020 by its authors (see COPYING, COPYING.LESSER, AUTHORS).
+ *
+ * The copyrights for code used from other parties are included in
+ * the corresponding files.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ *
+ * @END LICENSE
+ */
+
+#include <algorithm>
+#include <numeric>
+#include <regex>
+
+#include "helpers/combinatorial.h"
+#include "helpers/timer.h"
+#include "helpers/string_algorithms.h"
+
+#include "sparse_operator.h"
+
+namespace forte {
+
+void SparseOperator::add_term(const std::vector<std::tuple<bool, bool, int>>& op_list,
+                              double coefficient, bool allow_reordering) {
+    op_list_.push_back(SQOperator(op_list, coefficient, allow_reordering));
+}
+
+void SparseOperator::add_term(const SQOperator& sqop) { op_list_.push_back(sqop); }
+
+std::vector<std::tuple<bool, bool, int>> sparse_parse_ops(const std::string& s) {
+    // reverse the operator order
+    auto clean_s = s.substr(1, s.size() - 2);
+
+    auto ops_str = split_string(clean_s, " ");
+    std::reverse(ops_str.begin(), ops_str.end());
+
+    std::vector<std::tuple<bool, bool, int>> ops_vec_tuple;
+    for (auto op_str : ops_str) {
+        size_t len = op_str.size();
+        bool creation = op_str[len - 1] == '+' ? true : false;
+        bool alpha = op_str[len - 2] == 'a' ? true : false;
+        int orb = stoi(op_str.substr(0, len - 2));
+        ops_vec_tuple.push_back(std::make_tuple(creation, alpha, orb));
+    }
+    return ops_vec_tuple;
+}
+
+void SparseOperator::add_term_from_str(std::string str, double coefficient, bool allow_reordering) {
+    // the regex to parse the entries
+    std::regex re("\\s*(\\[[0-9ab\\+\\-\\s]*\\])");
+    // the match object
+    std::smatch m;
+
+    // here we match onethe terms of the form [<orb><a/b><+/-> ...]
```

Comment:
This comment isn't clear to me. I think you mean that you match terms of that form, translate them into a term, and add that to the operator. Things _not_ of that form raise an error.